File: letsPollApi/src/main/java/com/serverless/letspoll/handlers/RespondentRegistrar.java
package com.serverless.letspoll.handlers;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.serverless.letspoll.commons.DatabaseAccessUtils;
import com.serverless.letspoll.commons.RandomIdGenerator;
import com.serverless.letspoll.models.ApiGatewayResponse;
import com.serverless.letspoll.models.Poll;
import com.serverless.letspoll.models.generated.Tables;
import com.serverless.letspoll.models.requests.RespondentRegisterationRequest;
import com.serverless.letspoll.models.responses.RespondentRegistrationResponse;
import org.jooq.DSLContext;
import org.jooq.impl.DSL;

import java.util.Collections;
import java.util.List;
import java.util.Map;


public class RespondentRegistrar implements RequestHandler<Map<String,Object>, ApiGatewayResponse> {

    @Override public ApiGatewayResponse handleRequest(Map<String, Object> input, Context context) {

        final ObjectMapper mapper = new ObjectMapper();
        final RespondentRegisterationRequest respondentRegisterationRequest = mapper.convertValue(input, RespondentRegisterationRequest.class);
        String respondentId = RandomIdGenerator.getRandomString("RID");
        DSLContext dslContext = DatabaseAccessUtils.getDatabaseConnection();
        try{
          dslContext.insertInto(Tables.RESPONDENT,Tables.RESPONDENT.RESPONDENT_ID,Tables.RESPONDENT.RESPONDENT_EMAIL_ID,Tables.RESPONDENT.RESPONDENT_TOKEN,Tables.RESPONDENT.RESPONDENT_DISPLAY_NAME)
                .values(respondentId,respondentRegisterationRequest.getEmailId(),respondentRegisterationRequest.getToken(),respondentRegisterationRequest.getEmailId())
                .execute();
        }catch (Exception e){
            e.printStackTrace();
            RespondentRegistrationResponse response = new RespondentRegistrationResponse();
            response.setLetsPollRespondentId(respondentId);
            return ApiGatewayResponse.builder().setStatusCode(409)
                .setObjectBody("Could not create the respondent")
                .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                .build();
        }

        RespondentRegistrationResponse response = new RespondentRegistrationResponse();
        response.setLetsPollRespondentId(respondentId);
        return ApiGatewayResponse.builder().setStatusCode(200)
                .setObjectBody(response)
                .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                .build();
        }


}

File: letsPollApi/src/main/kotlin/com/serverless/letspoll/handlers/RespondentRegistrar.kt
package com.serverless.letspoll.handlers

import com.amazonaws.services.lambda.runtime.Context
import com.amazonaws.services.lambda.runtime.RequestHandler
import com.fasterxml.jackson.databind.ObjectMapper
import com.serverless.letspoll.commons.DatabaseAccessUtils
import com.serverless.letspoll.commons.getRandomString
import com.serverless.letspoll.models.ApiGatewayResponse
import com.serverless.letspoll.models.generated.Tables
import com.serverless.letspoll.models.generated.tables.records.RespondentRecord
import com.serverless.letspoll.models.requests.RespondentRegisterationRequest
import com.serverless.letspoll.models.responses.RespondentRegistrationResponse
import java.io.IOException
import java.util.*


class RespondentRegistrar : RequestHandler<Map<String, Any>, ApiGatewayResponse> {

    override fun handleRequest(input: Map<String, Any>, context: Context): ApiGatewayResponse {
        val respondentRegisterationRequest: RespondentRegisterationRequest
        try {
            respondentRegisterationRequest = ObjectMapper().readValue(input["body"] as String, RespondentRegisterationRequest::class.java)
        } catch (e: IOException) {
            e.printStackTrace()
            return ApiGatewayResponse.builder().setStatusCode(400)
                    .setObjectBody("could not parse the request body properly")
                    .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                    .build()

        }

        val respondentId = getRandomString("RID")
        val dslContext = DatabaseAccessUtils.getDatabaseConnection()
        dslContext?.let {
            try {
                it.insertInto<RespondentRecord, String, String, String, String>(Tables.RESPONDENT, Tables.RESPONDENT.RESPONDENT_ID,
                        Tables.RESPONDENT.RESPONDENT_EMAIL_ID, Tables.RESPONDENT.RESPONDENT_TOKEN,
                        Tables.RESPONDENT.RESPONDENT_DISPLAY_NAME)
                        .values(respondentId, respondentRegisterationRequest.emailId,
                                respondentRegisterationRequest.token,
                                respondentRegisterationRequest.displayName).execute()
                return ApiGatewayResponse.builder().setStatusCode(200).setObjectBody(RespondentRegistrationResponse(respondentId))
                        .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                        .build()
            } catch (e: Exception) {
                e.printStackTrace()
                return ApiGatewayResponse.builder().setStatusCode(409)
                        .setObjectBody("Could not create the respondent")
                        .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                        .build()
            }
        } ?: kotlin.run {
            return ApiGatewayResponse.builder().setStatusCode(409)
                    .setObjectBody("Could not create the respondent")
                    .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                    .build()
        }
    }

}

